File: 496-next-greater-element-i/496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        nextGreater = {}
        stack = []
        for num in nums2:
            while stack and stack[-1] < num:
                nextGreater[stack.pop()] = num
            stack.append(num)
        
        while stack:
            nextGreater[stack.pop()] = -1
        
        ans = []
        for num in nums1:
            ans.append(nextGreater[num])
        return ans

File: 496-next-greater-element-i/496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Stack<Integer> stack = new Stack<>();
        Map<Integer, Integer> nextGreater = new HashMap<>();
        for(int i = 0; i < nums2.length; i++){
            while(stack.size() > 0 && stack.peek() < nums2[i]){
                nextGreater.put(stack.pop(), nums2[i]);
            }
            stack.push(nums2[i]);
        }
        
        int[] ans = new int[nums1.length];
        for(int i = 0; i < nums1.length; i++){
            int greater = nextGreater.getOrDefault(nums1[i], -1);
            ans[i] = greater;
        }
        return ans;
    }
}